ext/reflection/php_reflection.c

```diff
@@ -6116,33 +6116,32 @@ static zend_bool is_ignorable_reference(HashTable *ht, zval *ref) {
 	return Z_TYPE_P(Z_REFVAL_P(ref)) != IS_ARRAY || Z_ARRVAL_P(Z_REFVAL_P(ref)) != ht;
 }
 
-/* {{{ proto public ReflectionReference|null ReflectionReference::fromArrayElement(array array, mixed key)
+/* {{{ proto public ReflectionReference|null ReflectionReference::fromArrayElement(array array, string|int key)
  *     Create ReflectionReference for array item. Returns null if not a reference. */
 ZEND_METHOD(ReflectionReference, fromArrayElement)
 {
-	HashTable *ht;
-	zval *key, *item;
+	zval *item, *array;
+	zend_string *string_key = NULL;
+	zend_long int_key = 0;
 	reflection_object *intern;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "hz", &ht, &key) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_ARRAY(array)
+		Z_PARAM_STR_OR_LONG(string_key, int_key)
```

Comment:
I think in case of array keys, we usually follow "array key coercion": https://github.com/php/php-src/pull/4887#issuecomment-549328298

Reply:
@kocsismate Sorry for the delay in replying to you. Mind to guide me what do I need to do? I read the PR you linked but couldn't understand 😕

Reply:
@carusogabriel This switch is what you need to look at: https://github.com/php/php-src/pull/4887/files#diff-497f073aa1ab88afcb8b248fc25d2a12R6346

This is necessary because type coercion works a bit differently in case of arrays than usually. If I'm not mistaken, the differences are the most notable when using strict types. If `Z_PARAM_STR_OR_LONG ` is used in strict mode, then for example a `false` argument will result in a `TypeError`, however, the key in `$array[false] = 1` will coerce into `0`.

This is the main reason `array_key_exists()` was implemented by respecting these rules. As far as I can see, `ReflectionReference::fromArrayElement()` could also benefit from it, too.

P.S: `array_column()` uses `Z_PARAM_STR_OR_LONG_OR_NULL`, so that might be another candidate to change.

Reply:
Hm... I'm a bit unsure on this one. I think there might be value in using `int|string`, specifically because it is stricter under strict_types. Under weak types it will have the same behavior as array keys with these exceptions:

 * `resource` will throw instead of warning and converting to int. I think that's okay.
 * `null` will be interpreted as `0` instead of `""`. This is actually unspecified under union types, because they only specify the userland behavior where null is always rejected. This behavior happened as an accident of the implementation -- maybe we should make this case be throwing instead, for zpp union types.

Reply:
The reason why I think a less strict check would be advantageous in this case is because I believe most people are not aware of the actual array type coercion rules (at least I myself wasn't until I started working on php-src), that's why I can imagine that people would expect that if they added a `false` key to an array, then they can also provide `false` as every function and method parameter which deals with array keys.

Or doesn't this argumentation make much sense?

My main motivation is to have consistent type checks for array keys, so I'm ok if you say that usual ZPP checks should be used here (and the plan is to do the same in the other related areas).

Reply:
Bumping this one. :) Apparently, As my impression was, even people who follow the internal discussions/changes are not very familiar with array coercion rules (e.g. https://externals.io/message/111566), so the union type behaviour in strict mode would be surely surprising for many people. The behaviour differences in weak mode compared to normal array coercion rules seem completely ok (I like the idea of throwing for `null`).

So... it seems ok going either way :) Maybe a litmus test could be to evaluate what we want to do with array
coercion rules on the long run. If it's desirable to make them stricter, then let's definitely go with the union mode, otherwise the method used in `array_key_exists()` could also be used.

Reply:
And FYI, `array_column()` uses the string|int ZPP macro :)

Reply:
I'll try to finish it this weekend, thanks for the reminder!

Reply:
Given we have other precedent for `string|int` ZPP for checking array keys, then most probably only a rebase is needed, if Nikita also agrees :)

Reply:
Yeah I'm okay with the string|int usage here.